File: 程序员面试金典/8.4幂集/subsets_3.java
// 方法3：二进制数
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> subsets = new ArrayList<>();
        for (int x = 0; x < Math.pow(2, nums.length); ++x)
            subsets.add(intToSubset(nums, x));
        return subsets;
    }

    private List<Integer> intToSubset(int[] nums, int x) {
        List<Integer> subset = new ArrayList<>();
        for (int i = 0; x != 0; x >>= 1, ++i)
            if ((x & 1) == 1)
                subset.add(nums[i]);
        return subset;
    }

}

File: 程序员面试金典/8.4幂集/subsets_3.py
# 方法3：二进制数
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        return [int2subset(nums, x) for x in range(2 ** len(nums))]


def int2subset(nums, x):
    subset = []
    i = 0
    while x:
        if x & 1:
            subset.append(nums[i])
        i += 1
        x >>= 1
    return subset
